include/deal.II/grid/grid_tools.h

```diff
@@ -1285,6 +1296,93 @@ namespace GridTools
+     * be used to reinit Utilities::MPI::RemotePointEvaluation.
+     */
+    template <int dim, int spacedim>
+    struct DistributedComputeIntersectionLocationsInternal
+    {
+      /**
+       * Intersection answers are assumed to be simplices (e.g. provided by
+       * CGAL)
+       */
+      using IntersectionAnswerType =
+        std::array<dealii::Point<spacedim>, dim + 1>;
+
+      /**
+       * Information of each intersection on sending/evaluation side. The
+       * elements of the tuple are as follows: 0) cell level and index, 1) rank
+       * of the owning process, 2) local index of the owning process, 3)
+       * found intersection, 4) requested intersection.
+       *
+       * @note The vector is sorted according to 1), 2).
+       */
+      std::vector<std::tuple<std::pair<int, int>,
+                             unsigned int,
+                             unsigned int,
+                             IntersectionAnswerType>>
+        send_components;
+
+      /**
+       * Ranks to send to.
+       */
+      std::vector<unsigned int> send_ranks;
+
+      /**
+       * Information of each received data value. The elements of the tuple are
+       * as follows: 0) rank of sender, 1) local index, 2) found intersections.
+       *
+       * @note The vector is sorted according to 1), 0), 2).
+       *
+       * @note Multiple intersections between cells can be found
+       */
+      std::vector<
+        std::tuple<unsigned int, unsigned int, IntersectionAnswerType>>
+        recv_components;
+
+      /**
+       * Pointers of ranges to found intersections for requested intersection.
+       */
+      std::vector<unsigned int> recv_ptrs;
+
+      /**
+       * Ranks from where data is received.
+       */
+      std::vector<unsigned int> recv_ranks;
+
+      /**
+       * Distribute quadrature points to found intersections and construct
+       * GridTools::internal::DistributedComputePointLocationsInternal from
+       * class members. This can be done without searching for points again
+       * since all information is locally known.
+       */
+      GridTools::internal::DistributedComputePointLocationsInternal<dim,
+                                                                    spacedim>
+      convert_to_distributed_compute_point_locations_internal(
+        const unsigned int                  n_quadrature_pnts,
+        const Triangulation<dim, spacedim> &tria,
+        const Mapping<dim, spacedim> &      mapping) const;
+
+    private:
+      /**
+       * Helper function for
+       * convert_to_distributed_compute_point_locations_internal(). It sends the
+       * indices associated to quadrature points at the receiver side to the
+       * sender side, where the information is needed to build
+       * GridTools::internal::DistributedComputePointLocationsInternal::send_components
+       */
+      std::map<unsigned int, std::vector<unsigned int>>
+      communicate_indices(
```

Comment:
I understand why you need this. You want to uniquely enumerate the points on receiver side. Put it differently, you would like to enumerate each intersection on receiver side; the point index is given by `intersection_index * q_points.size() + point_index`.

`DistributedComputePointLocationsInternal` stores this information. But actually RPE does not use this information, see: https://github.com/dealii/dealii/blob/3606a8e48154aa8eee98d5cf470069e341de7192/source/base/mpi_remote_point_evaluation.cc#L181-L193

My suggestion, would be to simply not enumerate the intersections/points.

Alternatively,  I would make the function optional and simplify the logic. 1) enumerate intersections; 2) sort according to ranks; 3) isend/irecv (each working on a single vector); 4) use the information with the assumption that `send_components` is sorted according to ranks.

Reply:
I have to check again, but when I tried without communicating the indices I got wrong results even in easy cases.

Reply:
I replace on master the following lines:
```diff
diff --git a/source/grid/grid_tools.cc b/source/grid/grid_tools.cc
index 8afe925..41bea70 100644
--- a/source/grid/grid_tools.cc
+++ b/source/grid/grid_tools.cc
@@ -6367,6 +6367,8 @@ namespace GridTools
         return request;
       };
 
+      unsigned int index = 0;
+
       const auto answer_request =
         [&](const unsigned int &other_rank,
             const RequestType & request) -> AnswerType {
@@ -6409,7 +6411,7 @@ namespace GridTools
                     send_components.emplace_back(
                       std::pair<int, int>(cell->level(), cell->index()),
                       other_rank,
-                      index_and_point.first,
+                      index++,
                       reference_position,
                       index_and_point.second,
                       numbers::invalid_unsigned_int);
```

... and it still works. The trick is to keep the order correctly, which should be the case also in your case.

Reply:
Sounds great :)

Reply:
I can confirm this :) I don't want to purge the code because it helped a lot during debugging and might be useful in the future, but I made it optional within [ee0907c](https://github.com/dealii/dealii/pull/15165/commits/ee0907c7f3247f2f3649ccb0781ccba5ebed76cb).